autopilot/contractor/contractor.go

```diff
@@ -1560,37 +889,535 @@ func renterFundsToExpectedStorage(renterFunds types.Currency, duration uint64, p
+		if err := bus.UpdateHostCheck(ctx, ctx.ApID(), h.host.PublicKey, *hc); err != nil {
+			return fmt.Errorf("failed to update host check for host %v: %w", h.host.PublicKey, err)
+		}
+		usabilityBreakdown.track(hc.Usability)
+
+		if !hc.Usability.IsUsable() {
+			logger.With("hostKey", h.host.PublicKey).
+				With("reasons", strings.Join(hc.Usability.UnusableReasons(), ",")).
+				Debug("host is not usable")
+		}
+	}
+
+	logger.Infow("host checks completed", usabilityBreakdown.keysAndValues()...)
+	return nil
+}
+
+func performPostMaintenanceTasks(ctx *mCtx, bus Bus, w Worker, alerter alerts.Alerter, cc contractChecker, rb revisionBroadcaster) error {
+	// fetch some contract and host info
+	allContracts, err := bus.Contracts(ctx, api.ContractsOpts{})
+	if err != nil {
+		return fmt.Errorf("failed to fetch all contracts: %w", err)
+	}
+	setContracts, err := bus.Contracts(ctx, api.ContractsOpts{ContractSet: ctx.ContractSet()})
+	if err != nil {
+		return fmt.Errorf("failed to fetch contracts: %w", err)
+	}
+	allHosts, err := bus.SearchHosts(ctx, api.SearchHostOptions{
+		Limit:         -1,
+		FilterMode:    api.HostFilterModeAllowed,
+		UsabilityMode: api.UsabilityFilterModeAll,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to fetch all hosts: %w", err)
+	}
+	usedHosts := make(map[types.PublicKey]struct{})
+	for _, c := range allContracts {
+		usedHosts[c.HostKey] = struct{}{}
+	}
+
+	// run revision broadcast on contracts in the new set
+	rb.runRevisionBroadcast(ctx, w, setContracts)
+
+	// register alerts for used hosts with lost sectors
+	var toDismiss []types.Hash256
+	for _, h := range allHosts {
+		if _, used := usedHosts[h.PublicKey]; !used {
+			continue
+		} else if registerLostSectorsAlert(h.Interactions.LostSectors*rhpv2.SectorSize, h.StoredData) {
+			alerter.RegisterAlert(ctx, newLostSectorsAlert(h.PublicKey, h.Settings.Version, h.Settings.Release, h.Interactions.LostSectors))
+		} else {
+			toDismiss = append(toDismiss, alerts.IDForHost(alertLostSectorsID, h.PublicKey))
+		}
+	}
+	if len(toDismiss) > 0 {
+		alerter.DismissAlerts(ctx, toDismiss...)
+	}
+
+	// prune refresh failures
+	cc.pruneContractRefreshFailures(allContracts)
+	return nil
+}
+
+func performContractMaintenance(ctx *mCtx, alerter alerts.Alerter, bus Bus, churn *accumulatedChurn, w Worker, cc contractChecker, cr contractReviser, rb revisionBroadcaster, remaining types.Currency, logger *zap.SugaredLogger) (bool, error) {
+	logger = logger.Named("performContractMaintenance").
+		Named(hex.EncodeToString(frand.Bytes(16))). // uuid for this iteration
+		With("contractSet", ctx.ContractSet())
+
+	// check if we want to run maintenance
+	if reason, skip := canSkipContractMaintenance(ctx, ctx.ContractsConfig()); skip {
+		logger.With("reason", reason).Info("skipping contract maintenance")
+		return false, nil
+	}
+	logger.Infow("performing contract maintenance")
+
+	// STEP 1: perform host maintenance
+	if err := performHostChecks(ctx, bus, logger); err != nil {
```

Comment:
We can do that but I'd make that a dedicated PR since we would probably move the pruning code out of the scanner and the host check code out of the contractor and into a new entity.

Reply:
Ok, problem with host checks is the (annoying) redundant IP field. That's also why currently `runHostChecks` is not persisting host checks until later that extra info gets annotated at which point we can persist them. Think you changed that though but it's still kind of annoying hard coupling between those two systems.

Reply:
I like the idea of having an extra `host` component though but we should probably rename the scanner and have it to everything, scan/check/prune. Right in time too because I _just_ refactored it!